### ISSUE TITLE
cformRect and CBitEditView::OnImageBoardMask() and CCellForm::CreateHexMask()?

### ISSUE THREAD

**Issue by wsu-cb** [CyberBoardPBEM/cbwindows#130]
What is `CBitEditView::OnImageBoardMask()` supposed to do for `cformRect` boards?  There is code that looks like it is intended to blacken the right and bottom edges of the tile, but, at least on my system, the tile doesn't actually change.  I have verified that CB 3.10 behaves this way as well.

Also, the `CCellForm::CreateHexMask()` code looks like it was never intended to be called for `cformRect`, but it actually is called when loading a saved `cformRect` board.  I haven't looked at this in CB 3.10 code, but it is already true in tag `v3.5-prerelease-2`.

**Comment by DLLarson** [comment]
I'm looking into it...

**Comment by wsu-cb** [comment]
> It shouldn't be called for rectangular cells. It must be benign in effect since we've never discovered it before. But it's still not correct.
> 

Does this apply to both `CCellForm::CreateHexMask()` and `CBitEditView::OnImageBoardMask()`, or just `CCellForm::CreateHexMask()`?

**Comment by DLLarson** [comment]
`CCellForm::CreateHexMask()` shouldn't be called for anything non-hex-like. I think the method is fine as it is.

`CBitEditView::OnImageBoardMask()` should be fine if `CreateHexMask()` isn't called to create a bitmask in the first place. Then the call to `pcf.GetMask()` should return NULL.

....or am I missing something else?

-Dale

**Comment by wsu-cb** [comment]
> `CCellForm::CreateHexMask()` shouldn't be called for anything non-hex-like. I think the method is fine as it is.
> 
> `CBitEditView::OnImageBoardMask()` should be fine if `CreateHexMask()` isn't called to create a bitmask in the first place. Then the call to `pcf.GetMask()` would return NULL.
> 
> ....or am I missing something else?
> 
> -Dale

When `pcf.GetMask()` returns NULL, `CBitEditView::OnImageBoardMask()` executes code that looks like it is intended to blacken the right and bottom edges of the tile, but, at least on my system, the tile doesn't actually change. I have verified that CB 3.10 behaves this way as well, i.e., the tile doesn't change.

**Comment by DLLarson** [comment]
>When pcf.GetMask() returns NULL, CBitEditView::OnImageBoardMask() executes code that looks like it is intended to blacken the right and bottom edges of the tile, but, at least on my system, the tile doesn't actually change. I have verified that CB 3.10 behaves this way as well, i.e., the tile doesn't change.

Yes. And I'm sure I had a good reason to do it. It probably extended back to my first gamebox: Tactics 2 since it had square cells. But I don't remember why. 

If you apply a cell mask to a tile that is larger than the board cell that was used to create the mask, you will end up with two lines that delineate the location and area what would be displayed when placing the tile in that boards cell layer. It's an editing tool to know what will be actually rendered. 

![image](https://github.com/CyberBoardPBEM/cbwindows/assets/175455/4c1c59e6-6cce-4ed6-8855-99d929ecfdc1)

I just don't have a use case for this but that doesn't mean else wouldn't find a use.

I've attached a corrective patch that should clean this code up some.

[0001-Fix-improper-call-to-CreateHexMask-when-shapes-are-r.patch](https://github.com/user-attachments/files/16069918/0001-Fix-improper-call-to-CreateHexMask-when-shapes-are-r.patch)

-Dale

**Comment by wsu-cb** [comment]
Have you tried adding a cell mask from a rectangular board to an oversized tile?  I don't see any black on the tile after applying a mask in that case, neither in CB 3.10 nor after applying the patch in the previous msg.

**Comment by DLLarson** [comment]
I just did it here:

![image](https://github.com/CyberBoardPBEM/cbwindows/assets/175455/4a581a14-1653-4936-906b-d19b673eaf4e)

I created a new tile several pixels larger than the board's cell (I just selected the board's cell size and edited the value provided to be larger.) Then I applied the board mask and I got the two lines delineating the visible area shown when dropped on the cell layer.

-Dale

**Comment by wsu-cb** [comment]
> I created a new tile several pixels larger than the board's cell (I just selected the board's cell size and edited the provided values to be larger.) Then I applied the board mask and I got the two lines delineating the visible area shown when dropped on the cell layer.
> 
> > neither in CB 3.10 nor after applying the patch in the previous msg.
> 
> It won't work in CB 3.10 because that code doesn't have the patch so it's rendering a mask that is the same size as the tile area. It has no black areas. So it's basically a no-op.
> 

So that explains why I couldn't get CB 3.10 to do it, and I have retried this after applying the patch, and this time it worked.

> I also noticed a bug in the eyedropper tool when I made the example. When I clicked a pixel the proper sampled color is shown but it is quickly overwritten with another color (in my case brown) upon mouse button release. This bug also exists in the current wide release. If you don't see it then it must be another problem triggered by High-DPI support in Windows.
> 

I had already found and fixed this while working on the wx conversion.  The problem was that the `ButtonUp` handler conversion of the mouse msg point to the bitmap point was done incorrectly, so the color of the wrong pixel was picked up.

**Comment by DLLarson** [comment]
>I had already found and fixed this while working on the wx conversion. The problem was that the ButtonUp handler conversion of the mouse msg point to the bitmap point was done incorrectly, so the color of the wrong pixel was picked up.

So I'm assuming that's a fix that's coming with your next batch of porting mods.

-Dale

**Comment by wsu-cb** [comment]
> > I had already found and fixed this while working on the wx conversion. The problem was that the ButtonUp handler conversion of the mouse msg point to the bitmap point was done incorrectly, so the color of the wrong pixel was picked up.
> 
> So I'm assuming that fix is coming with your next batch of porting mods?
> 

Yes, the PR I just posted includes a fix for this as part of the wx conversion, but not as a separate commit since I only noticed the problem when testing the converted code.
